zcash_client_sqlite/src/lib.rs

```diff
@@ -489,18 +487,24 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                     let tx_row = wallet::put_tx_meta(wdb.conn.0, tx, block.height())?;
 
                     // Mark notes as spent and remove them from the scanning cache
-                    for spend in &tx.sapling_spends {
+                    for spend in tx.sapling_spends() {
                         wallet::sapling::mark_sapling_note_spent(wdb.conn.0, tx_row, spend.nf())?;
                     }
 
-                    for output in &tx.sapling_outputs {
+                    for output in tx.sapling_outputs() {
                         // Check whether this note was spent in a later block range that
                         // we previously scanned.
-                        let spent_in = wallet::query_nullifier_map::<_, Scope>(
-                            wdb.conn.0,
-                            ShieldedProtocol::Sapling,
-                            output.nf(),
-                        )?;
+                        let spent_in = output
+                            .nf()
+                            .map(|nf| {
+                                wallet::query_nullifier_map::<_, Scope>(
+                                    wdb.conn.0,
+                                    ShieldedProtocol::Sapling,
+                                    nf,
+                                )
+                            })
+                            .transpose()?
+                            .flatten();
```

Comment:
I think that the condition that you're describing here should be unreachable in this release, because in order to select a note for spending we require the ZIP 32 account ID, which in the current implementation of `zcash_client_sqlite` must correspond to a UFVK in the wallet, which in turn must correspond to the USK that is used to attempt to then evaluate the proposal.

It is the responsibility of an implementation of `InputSource` to exclude notes associated with an IVK-only account; the current implementation of `zcash_client_sqlite` behaves correctly here.

Reply:
It is, of course, entirely possible for a FVK to be imported and used for scanning, and then the corresponding USK to be provided at the time that the user goes to actually spend - we can _never_ tell whether a user actually has access to the USK ahead of time, since all that the wallet concerns itself with internally is viewing keys.